flink-core/src/main/java/org/apache/flink/core/fs/LimitedConnectionsFileSystem.java

```diff
@@ -0,0 +1,1097 @@
+
+		/** The number of bytes written the last time that the {@link #checkNewBytesAndMark(long)}
+		 * method was called. It is important to initialize this with {@code -1} so that the
+		 * first check (0 bytes) always appears to have made progress. */
+		private volatile long lastCheckBytes = -1;
+
+		/** The timestamp when the last inactivity evaluation was made. */
+		private volatile long lastCheckTimestampNanos;
+
+		/** Flag tracking whether the stream was already closed, for proper inactivity tracking. */
+		private AtomicBoolean closed = new AtomicBoolean();
+
+		OutStream(
+				FSDataOutputStream originalStream,
+				LimitedConnectionsFileSystem fs) {
+
+			this.originalStream = checkNotNull(originalStream);
+			this.fs = checkNotNull(fs);
+		}
+
+		// --- FSDataOutputStream API implementation
+
+		@Override
+		public void write(int b) throws IOException {
+			try {
+				originalStream.write(b);
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public void write(byte[] b, int off, int len) throws IOException {
+			try {
+				originalStream.write(b, off, len);
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public long getPos() throws IOException {
+			try {
+				return originalStream.getPos();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+				return -1; // silence the compiler
+			}
+		}
+
+		@Override
+		public void flush() throws IOException {
+			try {
+				originalStream.flush();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public void sync() throws IOException {
+			try {
+				originalStream.sync();
+			}
+			catch (IOException e) {
+				handleIOException(e);
+			}
+		}
+
+		@Override
+		public void close() throws IOException {
+			if (closed.compareAndSet(false, true)) {
+				try {
+					originalStream.close();
+				}
+				catch (IOException e) {
+					handleIOException(e);
+				}
+				finally {
+					fs.unregisterOutputStream(this);
+				}
+			}
+		}
+
+		@Override
+		public void closeDueToTimeout() throws IOException {
+			this.timeoutException = new StreamTimeoutException();
+			close();
+		}
+
+		@Override
+		public boolean isClosed() {
+			return closed.get();
+		}
+
+		@Override
+		public long getLastCheckTimestampNanos() {
+			return lastCheckTimestampNanos;
+		}
+
+		@Override
+		public boolean checkNewBytesAndMark(long timestamp) throws IOException {
+			// remember the time when checked
+			lastCheckTimestampNanos = timestamp;
+
+			final long bytesNow = originalStream.getPos();
+			if (bytesNow > lastCheckBytes) {
+				lastCheckBytes = bytesNow;
+				return true;
+			}
+			else {
+				return false;
+			}
+		}
+
+		private void handleIOException(IOException exception) throws IOException {
+			if (timeoutException == null) {
+				throw exception;
+			} else {
+				// throw a new exception to capture this call's stack trace
+				throw new StreamTimeoutException(timeoutException);
```

Comment:
It might make sense to also store the `IOException` inside the `StreamTimeoutException`, because this could lose some information e.g. about an exception that happens as part of closing the stream.

Reply:
Adding that exception as a suppressed exception...